include/onnxruntime/core/platform/ort_mutex.h

```diff
@@ -3,11 +3,103 @@
+
+// Q: Why std::mutex can't make it?
+// A: VC runtime has to support Windows XP at ABI level. But we don't have such requirement.
+
+// Q: Is OrtMutex faster than std::mutex?
+// A: Sure
+
+class OrtMutex {
+ private:
+  SRWLOCK data_ = SRWLOCK_INIT;
+
+ public:
+  constexpr OrtMutex() = default;
+  // SRW locks do not need to be explicitly destroyed.
+  ~OrtMutex() = default;
+  OrtMutex(const OrtMutex&) = delete;
+  OrtMutex& operator=(const OrtMutex&) = delete;
+  void lock() { AcquireSRWLockExclusive(native_handle()); }
+  bool try_lock() noexcept { return TryAcquireSRWLockExclusive(native_handle()) == TRUE; }
+  void unlock() noexcept { ReleaseSRWLockExclusive(native_handle()); }
+  using native_handle_type = SRWLOCK*;
+
+  __forceinline native_handle_type native_handle() { return &data_; }
+};
+
+class OrtCondVar {
+  CONDITION_VARIABLE native_cv_object = CONDITION_VARIABLE_INIT;
+
+ public:
+  constexpr OrtCondVar() noexcept = default;
+  ~OrtCondVar() = default;
+
+  OrtCondVar(const OrtCondVar&) = delete;
+  OrtCondVar& operator=(const OrtCondVar&) = delete;
+
+  void notify_one() noexcept { WakeConditionVariable(&native_cv_object); }
+  void notify_all() noexcept { WakeAllConditionVariable(&native_cv_object); }
+
+  void wait(std::unique_lock<OrtMutex>& lk) {
+    if (SleepConditionVariableSRW(&native_cv_object, lk.mutex()->native_handle(), INFINITE, 0) != TRUE) {
+      std::terminate();
+    }
+  }
+  template <class _Predicate>
+  void wait(std::unique_lock<OrtMutex>& __lk, _Predicate __pred);
+
+  /**
+   * returns cv_status::timeout if the wait terminates when Rel_time has elapsed. Otherwise, the method returns
+   * cv_status::no_timeout.
+   * @param cond_mutex A unique_lock<OrtMutex> object.
+   * @param rel_time A chrono::duration object that specifies the amount of time before the thread wakes up.
+   * @return returns cv_status::timeout if the wait terminates when Rel_time has elapsed. Otherwise, the method returns
+   * cv_status::no_timeout
+   */
+  template <class Rep, class Period>
+  std::cv_status wait_for(std::unique_lock<OrtMutex>& cond_mutex, const std::chrono::duration<Rep, Period>& rel_time);
+  using native_handle_type = CONDITION_VARIABLE*;
+
+  native_handle_type native_handle() { return &native_cv_object; }
+
+ private:
+  void timed_wait_impl(std::unique_lock<OrtMutex>& __lk,
+                       std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>);
+};
+
+template <class _Predicate>
+void OrtCondVar::wait(std::unique_lock<OrtMutex>& __lk, _Predicate __pred) {
+  while (!__pred()) wait(__lk);
+}
+
+template <class Rep, class Period>
+std::cv_status OrtCondVar::wait_for(std::unique_lock<OrtMutex>& cond_mutex,
+                                    const std::chrono::duration<Rep, Period>& rel_time) {
+  // TODO: is it possible to use nsync_from_time_point_ ?
+  using namespace std::chrono;
+  if (rel_time <= duration<Rep, Period>::zero())
+    return std::cv_status::timeout;
```

Comment:
Is it necessary to special case this? Seems like it would be an unusual scenario and we're paying the cost of this check for all usages.